File: elixir/lib/bonemesh/frame.ex
defmodule Bonemesh.Frame do
  @moduledoc """
  The BoneMesh v3 frame reader/writer (protocol.md §2): one newline-terminated
  UTF-8 JSON object per frame, within a hard size cap — the enforcement point
  for defect D7. Classification verdicts match the shared corpus
  (spec/corpus/framing.json), so a node in any language agrees on which frames
  are well formed.
  """

  @handshake_cap 32768
  @transport_cap 65536

  @doc "Maximum handshake frame size in bytes (including the newline)."
  def handshake_cap, do: @handshake_cap
  @doc "Maximum transport frame size in bytes (including the newline)."
  def transport_cap, do: @transport_cap

  @doc """
  Classifies the first frame in `raw` against `cap`. Returns `{:ok, object}` or
  `{:reject, reason}`, reading only up to the first newline.
  """
  @spec classify(binary(), pos_integer()) :: {:ok, map()} | {:reject, String.t()}
  def classify(raw, cap) do
    case :binary.match(raw, "\n") do
      :nomatch ->
        {:reject, "no-newline"}

      {nl, 1} ->
        cond do
          nl + 1 > cap -> {:reject, "oversize"}
          nl == 0 -> {:reject, "empty"}
          true -> classify_content(binary_part(raw, 0, nl))
        end
    end
  end

  @doc "Encodes an object as a frame body followed by a single newline."
  @spec encode(map()) :: binary()
  def encode(object), do: JSON.encode!(object) <> "\n"

  defp classify_content(content) do
    cond do
      not String.valid?(content) ->
        {:reject, "invalid-utf8"}

      binary_part(content, 0, 1) == "[" ->
        {:reject, "not-an-object"}

      binary_part(content, 0, 1) != "{" ->
        {:reject, "invalid-json"}

      true ->
        decode(content)
    end
  end

  defp decode(content) do
    case JSON.decode(content) do
      {:ok, map} when is_map(map) ->
        {:ok, map}

      {:ok, _other} ->
        {:reject, "not-an-object"}

      {:error, {:invalid_byte, pos, _byte}} ->
        # An invalid byte after a complete object is trailing data; otherwise the
        # object itself is malformed.
        case JSON.decode(binary_part(content, 0, pos)) do
          {:ok, map} when is_map(map) -> {:reject, "trailing-data"}
          _ -> {:reject, "invalid-json"}
        end

      {:error, _} ->
        {:reject, "invalid-json"}
    end
  end
end

File: python/bonemesh/frame.py
"""BoneMesh v3 frame reader/writer (protocol.md §2).

One newline-terminated UTF-8 JSON object per frame within a hard size cap (defect
D7). Classification verdicts match the shared corpus (spec/corpus/framing.json).

A frame body must be exactly one JSON value with nothing but whitespace after it:
``{...} X`` is ``trailing-data``, not ``invalid-json``. ``json.JSONDecoder``'s
``raw_decode`` gives the end of the first complete value, which keeps those two
verdicts distinct without a hand-rolled scanner.

Python's ``json`` accepts ``NaN``, ``Infinity`` and ``-Infinity`` by default,
which the corpus requires be rejected as ``invalid-json`` (strict RFC 8259).
``parse_constant`` is what turns them back into errors -- without it the five
``lenient-*`` cases would diverge from the other implementations.
"""

from __future__ import annotations

import json

HANDSHAKE_CAP = 32768
TRANSPORT_CAP = 65536

_WS = " \t\n\r"


def _reject_constant(name: str):
    raise ValueError(f"non-finite number not permitted: {name}")


_DECODER = json.JSONDecoder(parse_constant=_reject_constant)


def classify(raw: bytes, cap: int) -> tuple[dict | None, str | None]:
    """Returns ``(obj, None)`` on success or ``(None, reason)`` on rejection.

    Reads only up to the first newline.
    """
    nl = raw.find(b"\n")
    if nl < 0:
        return None, "no-newline"
    if nl + 1 > cap:
        return None, "oversize"
    content = raw[:nl]
    if len(content) == 0:
        return None, "empty"

    try:
        text = content.decode("utf-8")
    except UnicodeDecodeError:
        return None, "invalid-utf8"

    start = 0
    while start < len(text) and text[start] in _WS:
        start += 1
    try:
        value, end = _DECODER.raw_decode(text, start)
    except ValueError:
        return None, "invalid-json"

    rest = end
    while rest < len(text) and text[rest] in _WS:
        rest += 1
    if rest < len(text):
        return None, "trailing-data"

    if not isinstance(value, dict):
        return None, "not-an-object"
    return value, None


def encode(obj) -> bytes:
    """Encode an object as a frame body followed by a newline."""
    return (json.dumps(obj, separators=(",", ":"), ensure_ascii=False) + "\n").encode("utf-8")
